ipa-core/src/protocol/ipa_prf/boolean_ops/share_conversion_aby.rs

```diff
@@ -126,84 +128,113 @@ where
-        &x,
-    )
-    .await?;
+    assert!(NC * input_shares.len() < 120_000);
+    // get context
+    let m_ctx = dzkp_validator.context();
+    // generate sub contexts
+    let ctx_gen = &m_ctx.narrow(&Step::GenerateSecretSharing);
+    let ctx_masks = &m_ctx.narrow(&Step::IntegerAddBetweenMasks);
+    let ctx_mask_x = &m_ctx.narrow(&Step::IntegerAddMaskToX);
+
+    // vector of intermediate results
+    let mut vec_sh_y =
+        Vec::<BitDecomposed<AdditiveShare<Boolean, NC>>>::with_capacity(input_shares.len());
+    let mut vec_sh_r =
+        Vec::<BitDecomposed<AdditiveShare<Boolean, NC>>>::with_capacity(input_shares.len());
+    let mut vec_sh_s =
+        Vec::<BitDecomposed<AdditiveShare<Boolean, NC>>>::with_capacity(input_shares.len());
+
+    for (i, x) in input_shares.iter().enumerate() {
+        let record_id_x = first_record_id + i;
+        // generate sh_r = (0, 0, sh_r) and sh_s = (sh_s, 0, 0)
+        // the two highest bits are set to 0 to allow carries for two additions
+        let (sh_r, sh_s) = gen_sh_r_and_sh_s::<_, BITS, NC>(ctx_gen, record_id_x);
+
+        // addition r+s might cause carry,
+        // this is no problem since we have set bit 254 of sh_r and sh_s to 0
+        let sh_rs = {
+            let (mut rs_with_higherorderbits, _) =
+                integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
+                    ctx_masks.clone(),
+                    record_id_x,
+                    &sh_r,
+                    &sh_s,
+                )
+                .await?;
+
+            // PRSS/Multiply masks added random highest order bit,
+            // remove them to not cause overflow in second addition (which is mod 256):
+            rs_with_higherorderbits[BITS - 1] = AdditiveShare::<Boolean, NC>::ZERO;
+
+            // return rs
+            rs_with_higherorderbits
+        };
+
+        // addition x+rs, where rs=r+s might cause carry
+        // this is not a problem since bit 255 of rs is set to 0
+        vec_sh_y.push(
+            integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
+                ctx_mask_x.clone(),
+                record_id_x,
+                &sh_rs,
+                x,
+            )
+            .await?
+            .0,
+        );
+        vec_sh_s.push(sh_s);
+        vec_sh_r.push(sh_r);
+    }
 
     // validate before reveal
-    validator.validate().await?;
-
-    // this leaks information, but with negligible probability
-    let mut y = (m_ctx.role() != Role::H3).then(|| Vec::with_capacity(NC));
-    for i in 0..BITS {
-        let y_bit = partial_reveal(
-            m_ctx.narrow(&Step::RevealY(i)),
-            record_id,
-            Role::H3,
-            sh_y.get(i).unwrap(),
-        )
-        .await?;
-        match (&mut y, y_bit) {
-            (Some(y), Some(y_bit)) => y.push(y_bit),
-            (None, None) => (),
-            _ => unreachable!("inconsistent partial_reveal behavior"),
+    dzkp_validator.validate_chunk(ctx_counter).await?;
```

Comment:
It occurs to me that we should have the implementation of reveal for DZKP contexts do an is_validated check. I made a note and will plan to look at that.

Reply:
is there a better way to deal with chunks inside the validator that does not require seeding the context from the callee? Copying the specific step all the way down through all functions in MPC seems very wrong to me

Reply:
There was some discussion of this [here](https://github.com/private-attribution/ipa/pull/1145#discussion_r1647375213). I think there probably is a better way, but it requires a little rearranging of the internals of the validator/context. I am hoping to take a look at this at some point.